### PR TITLE
Add reference documentation for DurableObjectState

### PR DESCRIPTION
**Pull request by joshthoward**
### Summary

Add reference documentation for DurableObjectState. A lot of this is taken from the existing WebSockets documentation. I will shortly refactor those docs to operate more like a guide for using WebSockets with and without hibernation. 

### Screenshots (optional)

Note that the RHS bar needs some fixes to the CSS. Text is being wrapped when it shouldn't be. 

Screenshot 2024-10-10 at 12 30 18 PM

### Documentation checklist

- [ ] The [documentation style guide](https://developers.cloudflare.com/style-guide/) has been adhered to.
- [ ] If a larger change - such as adding a new page- an issue has been opened in relation to any incorrect or out of date information that this PR fixes.
- [ ] Files which have changed name or location have been allocated [redirects](https://developers.cloudflare.com/pages/configuration/redirects/#per-file).

### CODE REVIEW

File: src/content/docs/durable-objects/api/state.mdx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  order: 4
+---
+
+import { Tabs, TabItem } from "~/components";
+
+## Description
+
+The `DurableObjectState` interface is accessible as an instance property on the Durable Objects class. This interface encapsulates methods that modify the state of a Durable Object, for example which WebSockets are attached to a Durable Object instance or how the runtime should handle concurrent Durable Object requests.
+
+The `DurableObjectState` interface is different that the [Storage API](/durable-objects/api/storage-api) in that it does not have top level methods which manipulate persistant application data. These methods are instead encapsulated in the [`DurableObjectStorage`](/durable-objects/api/storage-api) interface and accessed by [`DurableObjectState::storage`](/durable-objects/api/state/#storage).
```
Comment (lambrospetrou) [inline src/content/docs/durable-objects/api/state.mdx]:
```suggestion
The `DurableObjectState` interface is different than the [Storage API](/durable-objects/api/storage-api) in that it does not have top level methods which manipulate persistant application data. These methods are instead encapsulated in the [`DurableObjectStorage`](/durable-objects/api/storage-api) interface and accessed by [`DurableObjectState::storage`](/durable-objects/api/state/#storage).
```

File: src/content/docs/durable-objects/api/state.mdx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+<Tabs> <TabItem label="JavaScript" icon="seti:javascript">
+
+```js
+import { DurableObject } from "cloudflare:workers";
+
+// Durable Object
+export class MyDurableObject extends DurableObject {
+  // DurableObjectState is accessible via the ctx instance property
+	constructor(ctx, env) {
```
Comment (lambrospetrou) [inline src/content/docs/durable-objects/api/state.mdx:24]: Why are we naming it `ctx` and not `state`? There is a mismatch I feel here between the name and the type.

File: src/content/docs/durable-objects/api/state.mdx [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+}
+```
+
+</TabItem> </Tabs>
+
+## Methods
+
+### `waitUntil`
+
+`waitUntil` waits until the promise which is passed as a parameter resolves and can extends an event context up to 30 seconds after the last client disconnects.
```
Comment (lambrospetrou) [inline src/content/docs/durable-objects/api/state.mdx]:
```suggestion
`waitUntil` waits until the promise which is passed as a parameter resolves and extends an event context up to 30 seconds after the last client disconnects.
```
Do people understand what an "event context" is? I don't really.

File: src/content/docs/durable-objects/api/state.mdx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+- A required `WebSocket`.
+
+#### Return values
+
+- An `Array<string>` of tags.
+
+### `abort`
+
+`abort` is used to forcibly reset a Durable Object instance. A JavaScript Error with the message passed as a parameter will be logged. This error is not able to be caught within the applciation code.
```
Comment (lambrospetrou) [inline src/content/docs/durable-objects/api/state.mdx]:
```suggestion
`abort` is used to forcibly reset a Durable Object instance. A JavaScript `Error` with the message passed as a parameter will be logged. This error is not able to be caught within the application code.
```

File: src/content/docs/durable-objects/api/state.mdx (hunk 3 above)
Comment (joshthoward) [reply src/content/docs/durable-objects/api/state.mdx]: It's effectively how long we keep the DO running in memory. Do you have an idea on how to better frame this?

File: src/content/docs/durable-objects/api/state.mdx (hunk 2 above)
Comment (joshthoward) [reply src/content/docs/durable-objects/api/state.mdx:24]: This is for consistency with WorkerEntrypoints. It's documented elsewhere, but that documentation should be moved here.

**Comment by cacheburner** [comment]
Looks like the page preview failed to build? https://github.com/cloudflare/cloudflare-docs/actions/runs/11279062020/job/31368913837?pr=17471

File: src/content/docs/durable-objects/api/state.mdx (hunk 1 above)
Comment (Oxyjun) [inline src/content/docs/durable-objects/api/state.mdx]:
```suggestion
The `DurableObjectState` interface is different from the [Storage API](/durable-objects/api/storage-api) in that it does not have top-level methods which manipulate persistent application data. These methods are instead encapsulated in the [`DurableObjectStorage`](/durable-objects/api/storage-api) interface and accessed by [`DurableObjectState::storage`](/durable-objects/api/state/#storage).
```

File: src/content/docs/durable-objects/api/state.mdx (hunk 3 above)
Comment (Oxyjun) [reply src/content/docs/durable-objects/api/state.mdx]: Let's add this to the DO glossary, and have it hover in-line. Could you help with the definition, in a way that users would understand?

image

Comment (Oxyjun) [reply src/content/docs/durable-objects/api/state.mdx]: Glossary definition has been updated, the hover text will display the correct definition.

**Review by Oxyjun — APPROVED** [review]

**Pull request merged**